src/category_theory/localization/construction.lean

```diff
@@ -0,0 +1,194 @@
+
+/-- The relations by which we take the quotient in order to get the localized category. -/
+inductive relations ⦃X Y : (paths (loc_quiver W))⦄ : (X ⟶ Y) → (X ⟶ Y) → Prop
+| id (φ : X ⟶ Y) :
+  relations (ψ₁' W (𝟙 X.obj) ≫ φ) φ
+| comp {Z : paths (loc_quiver W)} (f : X.obj ⟶ Z.obj) (g : Z.obj ⟶ Y.obj) :
+  relations (ψ₁' W (f ≫ g)) (ψ₁' W f ≫ ψ₁' W g)
+| Winv₁ {Z : paths (loc_quiver W)} (w : X.obj ⟶ Z.obj) (hw : arrow.mk w ∈ W) (φ : X ⟶ Y) :
+  relations (ψ₁' W w ≫ ψ₂' W w hw ≫ φ) φ
+| Winv₂ {Z : paths (loc_quiver W)} (w : Z.obj ⟶ Y.obj) (hw : arrow.mk w ∈ W) (φ : X ⟶ Y) :
+  relations (φ ≫ ψ₂' W w hw ≫ ψ₁' W w) φ
+
+end localization
+
+/-- The localized category obtained by formally inverting the morphisms in `W : arrow_class C` -/
+@[derive category, nolint has_inhabited_instance]
+def localization := category_theory.quotient (localization.relations W)
+
+namespace localization
+
+/-- The obvious functor `C ⥤ localization W` -/
+def Q : C ⥤ W.localization :=
+{ obj := λ X, (quotient.functor _).obj (paths.of.obj ⟨X⟩),
+  map := λ X Y f, (quotient.functor _).map (ψ₁ W f),
+  map_id' := λ X, by convert quotient.sound _ (@relations.id _ _ W (paths.of.obj ⟨X⟩) _ (𝟙 _)),
+  map_comp' := λ X Z Y f g, quotient.sound _
+    (@relations.comp _ _ W (paths.of.obj ⟨X⟩) (paths.of.obj ⟨Y⟩) (paths.of.obj ⟨Z⟩) f g), }
+
+/-- The isomorphism in `localization W` associated to a morphism `w` in W -/
+def Wiso {X Y : C} (w : X ⟶ Y) (hw : arrow.mk w ∈ W) : iso ((Q W).obj X) ((Q W).obj Y) :=
+{ hom := (Q W).map w,
+  inv := (quotient.functor _).map (paths.of.map (sum.inr ⟨w, hw⟩)),
+  hom_inv_id' := quotient.sound _ (@relations.Winv₁ _ _ _
+    (paths.of.obj ⟨X⟩) _ (paths.of.obj ⟨Y⟩) w hw (𝟙 _)),
+  inv_hom_id' := quotient.sound _ (@relations.Winv₂ _ _ _
+    (paths.of.obj ⟨Y⟩) (paths.of.obj ⟨Y⟩) (paths.of.obj ⟨X⟩) w hw (𝟙 _)), }
+
+/-- The isomorphism in `localization W` associated to a morphism `w : W` -/
+def Wiso' (w : W) : iso ((Q W).obj w.1.left) ((Q W).obj w.1.right) :=
+Wiso W w.1.hom (by { rw arrow.mk_eq w.1, exact w.2, })
+
+lemma W_is_inverted_by_Q : W.is_inverted_by (Q W) := λ w, is_iso.of_iso (Wiso' W w)
+
+variables {W}
+
+variables (G : C ⥤ D) (hG : W.is_inverted_by G)
+
+include G hG
+
+/-- The lifting of a functor to the path category of `loc_quiver W` -/
+@[simps]
+def lift_to_path_category : paths (loc_quiver W) ⥤ D :=
+Quiv.lift
+{ obj := λ X, G.obj X.obj,
+  map := λ X Y, begin
+    rintro (f|⟨g, hg⟩),
+    { exact G.map f, },
+    { haveI : is_iso (G.map g) := hG ⟨arrow.mk g, hg⟩,
+      exact inv (G.map g), },
+  end, }
+
+/-- The lifting of a functor `C ⥤ D` inverting `W` as a functor `localization W ⥤ D` -/
+@[simps]
+def lift : localization W ⥤ D :=
+quotient.lift (relations W) (lift_to_path_category G hG)
+begin
+  rintro ⟨X⟩ ⟨Y⟩ f₁ f₂ r,
+  rcases r with (_|⟨⟨Z⟩, f, g⟩|⟨⟨Z⟩, w, hw⟩|⟨⟨Z⟩, w, hw⟩),
+  tidy,
+end
+
+@[simp]
+lemma fac : Q W ⋙ lift G hG = G :=
```

Comment:
Are we sure that we want functor equality here as opposed to an isomorphism? (Similarly for  `uniq` below.)

Reply:
This is the way the localization of categories is defined in the book by Gabriel-Zisman or in Quillen's Homotopical Algebra.